### ISSUE TITLE
[CA-drvs] nix prints "skipping suspicious writable file" after installing each instantiated CA derivation

### ISSUE THREAD

**Issue by zseri** [NixOS/nix#4806]
**Describe the bug**

When installing the result of a content-addressable derivation, nix prints "skipping suspicious writable file" for most, if not all contained (in the result) regular files, at least.

**Steps To Reproduce**

Precondition: store auto-optimization must be enabled.
run nix on a CA derivation which outputs regular files.
nix (concretely: `optimize-store.cc`) then proceeds to produce warnings "skipping suspicious writable file" for each installed regular file.

**Expected behavior**

This shouldn't happen, especially because this also makes auto-optimization mostly unusable/worthless.

**`nix-env --version` output**

nix-env (Nix) 2.4pre20210506_db6ab75

**Additional context**

This bug might be a consequence of rewriting. Maybe files aren't correctly reset to non-writable...
"rewriting hashes" happens at: `src/libstore/build/local-derivation-goal.cc:2295`, the written files have afaik the writable file mode bit set (set at `src/libutil/archive.cc:313`).
Something at seems to be going wrong: `src/libstore/build/local-derivation-goal.cc:2454` (this should, as far as I interpret the comment above it, reset the file mode bits, but it somehow doesn't).
tree optimisation seems to be invoked at: `src/libstore/build/local-derivation-goal.cc:2538`.
"skipping suspicious writable file" happens at: `src/libstore/optimise-store.cc:129`

related to: #4513

**Comment by zseri** [comment]
Maybe the problem lies here (suspicious, because the comment makes it seem that this is another shared code path between #4513 and #4806):
https://github.com/NixOS/nix/blob/de9e43c2ea8ac8752d42660c17a7b683682ef38b/src/libstore/build/local-derivation-goal.cc#L2461-L2491

test case:
```nix
let pkgs = import <nixpkgs> {};
in derivation {
  name = "iss4806tc";
  builder = "${pkgs.bash}/bin/bash";
  __contentAddressed = true;
  outputHashAlgo = "sha256";
  outputHashMode = "recursive";
  args = ["-c" "${pkgs.coreutils}/bin/mkdir -p $out; echo ... > $out/00"];
  system = builtins.currentSystem;
}
```
strace of `nix-build [test case file]`: https://ytrizja.de/misc/5c2efba66889e6113cff4f811b00d151683a720586db45a6732a5ac3fa1b60e4.strace

Another prerequisite for reproducibility: NixOS or at least a multi-user nix install. the test case above doesn't trigger the issue on my single-user installation.

**Comment by zseri** [comment]
This may be a DUP of #4775, I'll test it...

**Comment by zseri** [comment]
nope, isn't a DUP...
```
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libpcre.so.1.2.11'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libgcc_s.so.1'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libpthread_nonshared.a'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libmpfr.so.6.0.2'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libelf.so.0.8.13'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libutil-2.27.so'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libnss_db-2.27.so'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libmpc.so.3.1.0'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libpcreposix.so.0.0.6'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libpthread-2.27.so'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libnss_nis-2.27.so'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libgmpxx.so.4.5.2'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libcidn-2.27.so'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libz.so.1.2.11'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libcrypt-2.27.so'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libbz2.so.1.0.6'
warning: skipping suspicious writable file '/nix/store/v4z0cr49w5hdwbwlhgrcy1ibkl8czis7-bootstrap-tools/lib/libc-2.27.so'
```
nix-env (Nix) 2.4pre20210514_de9e43c

**Comment by zseri** [comment]
ok, happens again.

**Comment by zseri** [comment]
according to recent tests of my, this appears to be fixed now (with `>= nix-2.4pre20210514_de9e43c`, after garbage collecting all the (partially corrupted) previous CA builds)
